crates/pessimistic-proof-core/src/aggchain_data/multisig.rs

```diff
@@ -42,55 +39,36 @@ impl MultiSignature {
         keccak256_combine([buf.as_slice()])
     }
 
-    /// Ensure that the signatures refer to unique and existing signer.
-    fn unique_signers(&self) -> Result<(), MultisigError> {
-        let total = self.expected_signers.len();
-        let mut seen = vec![false; total];
-
-        for &(idx, _) in &self.signatures {
-            if idx >= total {
-                return Err(MultisigError::OutOfBoundSignerIndex { idx, total });
-            }
-
-            if seen[idx] {
-                return Err(MultisigError::HasDuplicateSigner {
-                    signer: self.expected_signers[idx],
-                });
-            }
-
-            seen[idx] = true;
-        }
-
-        Ok(())
-    }
-
     /// Verify signatures and ensure they are all from the expected set.
     pub fn verify(&self, commitment: B256) -> Result<(), MultisigError> {
-        self.unique_signers()?;
-
-        let has_invalid_signature = self.signatures.iter().any(|&(signer_idx, signature)| {
-            self.expected_signers
-                .get(signer_idx)
-                .and_then(|expected| {
-                    signature
-                        .recover_address_from_prehash(&commitment)
-                        .ok()
-                        .map(|recovered| recovered != *expected)
-                })
-                .unwrap_or(true)
-        });
-
-        if has_invalid_signature {
-            return Err(MultisigError::HasInvalidSignature);
+        if self.signatures.len() > self.expected_signers.len() {
+            return Err(MultisigError::OutOfBoundSignerIndex {
+                idx: self.signatures.len() - 1,
+                total: self.expected_signers.len(),
+            });
         }
 
-        if self.signatures.len() < self.threshold {
+        let nb_signatures = self.signatures.iter().filter(|s| s.is_some()).count();
+        if nb_signatures < self.threshold {
             return Err(MultisigError::UnderThreshold {
-                got: self.signatures.len(),
+                got: nb_signatures,
                 expected: self.threshold,
             });
         }
 
+        for (idx, signature) in self.signatures.iter().enumerate() {
+            let Some(signature) = signature else {
+                continue; // No signature is a valid signature
+            };
+            let Ok(recovered) = signature.recover_address_from_prehash(&commitment) else {
+                return Err(MultisigError::HasInvalidSignature { idx }); // Failed to recover from prehash
+            };
+            let expected = self.expected_signers[idx];
+            if recovered != expected {
+                return Err(MultisigError::HasInvalidSignature { idx }); // Signer doesn't match
+            }
```

Comment:
can be useful for troubleshooting to have two different errors here

```suggestion
            let recovered = signature.recover_address_from_prehash(&commitment).map_err(|e| MultisigError::Recovery { idx, source: e })?;
            
            let expected = self.expected_signers[idx];
            if recovered != expected {
                return Err(MultisigError::HasInvalidSignature {
                     idx,
                     recovered,
                     expected
                }); // Signer doesn't match
            }
```

- possible reasons for `MultisigError::Recovery`:
   - ill-formed signature (not necessarily just wrong signer or wrong message, just ill-formed ecdsa)

- possible reasons for `HasInvalidSignature`
   - wrong signer or wrong commitment